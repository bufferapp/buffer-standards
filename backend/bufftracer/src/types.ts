export const TAG_NAME = 'api'
export const TAG_LIMIT = 200 // This is the character limit including all the characters of the string `key:value`
export const TAG_KEY_LENGTH = `${TAG_NAME}.fields10:`.length // This is the string length of the `key:` part
export const TAG_VALUE_LIMIT = TAG_LIMIT - TAG_KEY_LENGTH // This is the remaining character lenth we can use to set fields value
export const TAG_SEPARATOR = '_' // DD will join array of string using "_" as a separator

// We need several fields properties because DD APM tags are limited to 200 characters including the key and value length:
// This results in tag being cut and incomplete data for us because the tag won't contain all fields being queried
// api.fields:account.id_account.currentorganization_currentorganization.id_currentorganization.isonebufferorganization_currentorganization.name_currentorganization.role_currentorganization.billing_billi
export type APIMetadataFields = {
  fields?: Array<string>
  fields2?: Array<string>
  fields3?: Array<string>
  fields4?: Array<string>
  fields5?: Array<string>
  fields6?: Array<string>
  fields7?: Array<string>
  fields8?: Array<string>
  fields9?: Array<string>
  fields10?: Array<string>
}

export type APIRequestMetadata = {
  name: 'core-api' | string
  type: 'graphql' | 'rpc' | 'rest'
  client: 'buffertools-graphql-playground' | string
  edge: string
  deprecatedFields: Array<string>
  args: Array<string>
} & APIMetadataFields

type Request<P, B> = {
  headers: Record<string, string>
  params: P
  body: B
}
export type GraphQlRequest = Request<Record<string, never>, { query: string }>
export type RPCRequest = Request<
  { method?: string },
  { name?: string; args?: string | Record<string, unknown> }
>
