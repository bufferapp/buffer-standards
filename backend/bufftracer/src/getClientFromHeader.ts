import { GraphQlRequest, RPCRequest } from './types'

export const HEADER_CLIENT_ID = 'x-buffer-client-id'
export const HEADER_CLIENT_UNKNOWN = 'unknown'

export function getClientFromHeader(args: {
  req: RPCRequest | GraphQlRequest
}): string {
  const { req } = args

  return req.headers[HEADER_CLIENT_ID] || HEADER_CLIENT_UNKNOWN
}
