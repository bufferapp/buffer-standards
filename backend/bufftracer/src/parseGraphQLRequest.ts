import {
  FieldNode,
  FragmentDefinitionNode,
  FragmentSpreadNode,
  getOperationAST,
  InlineFragmentNode,
  parse,
  Kind,
  DefinitionNode,
  DocumentNode,
} from 'graphql'
import { SelectionNode, SelectionSetNode } from 'graphql/language/ast'
import { APIRequestMetadata, GraphQlRequest } from './types'
import { getClientFromHeader } from './getClientFromHeader'
import * as env from 'env-var'

export function parseGraphQLRequest(args: {
  req: GraphQlRequest
}): APIRequestMetadata | null {
  const { req } = args

  // Parse GraphQL query into an AST (Abstract Syntax Tree) = schema represented in nested object
  const ast = parse(req.body.query)

  // Retrieve the Operation sub-AST corresponding to the query/mutation
  const operationAST = getOperationAST(ast, null)
  if (!operationAST) return null

  const operationType = operationAST.operation.toLowerCase()

  // Retrieve any Fragment sub-ASTs corresponding to spread fragments
  const fragmentASTs = getFragmentASTs({ ast })

  // Parse fields starting from the operation AST
  const result = parseFieldsAndArguments({
    selectionSet: operationAST.selectionSet,
    parentFieldName: operationType,
    fragmentASTs,
  })

  const edge = result.fieldList.shift()?.replace('.', ' ') || 'unknown'

  return {
    name: env.get('DD_SERVICE_NAME').required().asString(),
    type: 'graphql',
    client: getClientFromHeader({ req }),
    edge,
    fields: result.fieldList,
    deprecatedFields: [],
    args: result.argumentList,
  }
}

export function getFragmentASTs(args: {
  ast: DocumentNode
}): Record<string, FragmentDefinitionNode> {
  const result: Record<string, FragmentDefinitionNode> = {}
  for (const definition of args.ast.definitions) {
    if (isFragmentDefinitionNode(definition)) {
      result[definition.name.value] = definition
    }
  }
  return result
}

/**
 * Recursive function trasversing the AST tree and returning a list of fields and arguments
 *  - selectionSet: the current node of the tree section
 *  - parentFieldName: name of the parent field to compose field like 'billing.subscription'
 *  - fragmentASTs: the ASTs corresdonding to spread fragment. As opposed to inline fragment, spread fragment are representend in the operation AST as a leaf.
 *                  So, in order to parse these as fields too, we are going to complete the tree ourselves when we reach a spread fragment leaf.
 */
function parseFieldsAndArguments(args: {
  selectionSet: SelectionSetNode
  parentFieldName: string
  fragmentASTs: Record<string, FragmentDefinitionNode>
}): { fieldList: Array<string>; argumentList: Array<string> } {
  const { selectionSet, parentFieldName, fragmentASTs } = args

  let fieldList: Array<string> = []
  let argumentList: Array<string> = []

  // Else, loop over each child node of the current node
  for (const selectionNode of selectionSet.selections) {
    let childrenResult: {
      fieldList: Array<string>
      argumentList: Array<string>
    }

    if (isFieldNode(selectionNode)) {
      // Get name of current node and add it to the list
      const fieldName = selectionNode.name.value
      fieldList.push(`${parentFieldName}.${fieldName}`)

      // Get the arguments of the node if any
      if (selectionNode.arguments) {
        for (const fieldArgument of selectionNode.arguments) {
          argumentList.push(
            `${parentFieldName}.${fieldName}.${fieldArgument.name.value}`,
          )
        }
      }

      // Continue with child elements
      if (selectionNode.selectionSet) {
        childrenResult = parseFieldsAndArguments({
          selectionSet: selectionNode.selectionSet,
          parentFieldName: fieldName,
          fragmentASTs,
        })
        fieldList = fieldList.concat(childrenResult.fieldList)
        argumentList = argumentList.concat(childrenResult.argumentList)
      }
    } else if (isInlineFragmentNode(selectionNode)) {
      // Get name of current node and add it to the list
      // i.e. billing.OBBilling
      const fieldName = selectionNode.typeCondition?.name.value
      fieldList.push(`${parentFieldName}.${fieldName}`)

      // Continue with child elements
      childrenResult = parseFieldsAndArguments({
        selectionSet: selectionNode.selectionSet,
        parentFieldName: parentFieldName, // Here we don't want to pass the fragment name, we want to produce 'billing.subscription' and not 'OBBilling.subscription'
        fragmentASTs,
      })
      fieldList = fieldList.concat(childrenResult.fieldList)
      argumentList = argumentList.concat(childrenResult.argumentList)
    } else if (isFragmentSpreadNode(selectionNode)) {
      // Get name of the spread fragment and add it to the list
      // i.e. organization.ChannelsFields
      const fieldName = selectionNode.name.value
      fieldList.push(`${parentFieldName}.${fieldName}`)

      // We reached a leaf of the opertion AST that is a fragment node
      // If we have an AST for the spread fragment, let's use to keep on going
      if (fragmentASTs[fieldName]) {
        childrenResult = parseFieldsAndArguments({
          selectionSet: fragmentASTs[fieldName].selectionSet,
          parentFieldName: parentFieldName, // Here we don't want to pass the fragment name, we want to produce 'organization.channels' and not 'ChannelsFields.channels'
          fragmentASTs,
        })
        fieldList = fieldList.concat(childrenResult.fieldList)
        argumentList = argumentList.concat(childrenResult.argumentList)
      }
    }
  }
  return { fieldList, argumentList }
}

function isFieldNode(selectionNode: SelectionNode): selectionNode is FieldNode {
  return selectionNode.kind === Kind.FIELD
}

function isInlineFragmentNode(
  selectionNode: SelectionNode,
): selectionNode is InlineFragmentNode {
  return selectionNode.kind === Kind.INLINE_FRAGMENT
}

function isFragmentSpreadNode(
  selectionNode: SelectionNode,
): selectionNode is FragmentSpreadNode {
  return selectionNode.kind === Kind.FRAGMENT_SPREAD
}

function isFragmentDefinitionNode(
  definitionNode: DefinitionNode,
): definitionNode is FragmentDefinitionNode {
  return definitionNode.kind === Kind.FRAGMENT_DEFINITION
}
