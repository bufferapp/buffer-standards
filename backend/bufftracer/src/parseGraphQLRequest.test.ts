import { getFragmentASTs, parseGraphQLRequest } from './parseGraphQLRequest'
import { GraphQlRequest } from './types'
import { parse } from 'graphql'
import { HEADER_CLIENT_ID } from './getClientFromHeader'

export const query = `query accountGetAccount {
  account(id: "some-id") {
    email
    currentOrganization {
      billing {
        canAccessEngagement
        gateway {
          gatewayId
        }
        ... on MPBilling {
          subscriptions {
            product
            plan
          }
        }
        ...OBSubscriptionFields
      }
    }
    organizations {
      name
      ...ChannelsFields
    }
  }
}

fragment OBSubscriptionFields on OBBilling {
  subscription {
    plan {
      id
      name
    }
    quantity
  }
}

fragment ChannelsFields on AccountOrganization {
  channels(product: engage) {
    id
  }
}`

describe('getFragmentASTs', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('returns 2 fragment definition nodes', () => {
    // Arrange
    const ast = parse(query)

    // Act
    const result = getFragmentASTs({ ast })

    // Assert
    expect(Object.keys(result)).toHaveLength(2)
  })
})

describe('parseGraphQLRequest', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('returns fields and arguments for nodes, inline fragments and spread fragments', () => {
    // Arrange
    const req: GraphQlRequest = {
      headers: {
        [HEADER_CLIENT_ID]: 'webapp-account',
      },
      params: {},
      body: {
        query,
      },
    }

    // Act
    const result = parseGraphQLRequest({ req })

    // Assert
    expect(result).not.toBeNull()
    expect(result).toMatchSnapshot()
  })
})
