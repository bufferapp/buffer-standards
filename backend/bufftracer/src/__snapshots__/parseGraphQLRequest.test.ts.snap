// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseGraphQLRequest returns fields and arguments for nodes, inline fragments and spread fragments 1`] = `
Object {
  "args": Array [
    "query.account.id",
    "organizations.channels.product",
  ],
  "client": "webapp-account",
  "deprecatedFields": Array [],
  "edge": "query account",
  "fields": Array [
    "account.email",
    "account.currentOrganization",
    "currentOrganization.billing",
    "billing.canAccessEngagement",
    "billing.gateway",
    "gateway.gatewayId",
    "billing.MPBilling",
    "billing.subscriptions",
    "subscriptions.product",
    "subscriptions.plan",
    "billing.OBSubscriptionFields",
    "billing.subscription",
    "subscription.plan",
    "plan.id",
    "plan.name",
    "subscription.quantity",
    "account.organizations",
    "organizations.name",
    "organizations.ChannelsFields",
    "organizations.channels",
    "channels.id",
  ],
  "name": "foo-api",
  "type": "graphql",
}
`;
